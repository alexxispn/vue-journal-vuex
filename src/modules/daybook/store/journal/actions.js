import journalApi from "@/api/journalApi";

export const loadEntries = async ({commit}) => {
    const {data} = await journalApi.get("/entries.json")
    if (!data) {
        commit("setEntries", []);
    } else {
        const entries = [];
        for (let id of Object.keys(data)) {
            entries.push({
                id,
                ...data[id]
            });
        }
        commit("setEntries", entries);
        return entries;
    }
}

export const updateEntry = async ({commit}, entry) => {
    const newEntry = {
        ...entry
    }
    delete newEntry.id;
    await journalApi.put(`/entries/${entry.id}.json`, newEntry);
    commit("updateEntry", newEntry);
    return entry.id;
}

export const addEntry = async ({commit}, entry) => {
    const {data} = await journalApi.post("/entries.json", entry);
    commit("addEntry", {...entry, id: data.name});
    return data.name;
}


export const deleteEntry = async ({commit}, id) => {
    await journalApi.delete(`/entries/${id}.json`)
    commit("deleteEntry", id);
    return null
}
