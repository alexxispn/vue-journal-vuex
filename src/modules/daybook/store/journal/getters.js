export const getEntriesByTerm = (state) => (term) => {
    return state.entries.filter((entry) => {
        return entry.text.toLowerCase().includes(term.toLowerCase()) || entry.date.includes(term)
    })
}

export const getEntryById = (state) => (id) => {
    const entry = state.entries.find((entry) => entry.id === id)
    if (entry) {
        return entry
    }
    return null
}
