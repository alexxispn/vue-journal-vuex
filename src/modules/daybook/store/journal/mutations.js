export const setEntries = (state, entries) => {
    state.entries = entries;
}

export const updateEntry = (state, entry) => {
    state.entries = state.entries.map(e => e.id === entry.id ? entry : e);
}
