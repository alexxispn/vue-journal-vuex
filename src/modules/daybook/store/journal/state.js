export default () => ({
    isLoading: true,
    entries: [
        {
            id: '1',
            date: new Date().toDateString(),
            text: 'Lorem ipsum dolor sit amet consectetur adipisicing elit. Quisquam, quodr adipisicing elit. Quisquam, quodr adipisicing elit. Quisquam, quodr adipisicing elit. Quisquam, quod.',
            picture: null
        },
        {
            id: '2',
            date: new Date().toDateString(),
            text: 'Ipsum dolor sit amet consectetur adipisicing elit. Quisquam, quodr adipisicing elit. Quisquam, quodr adipisicing elit. Quisquam, quod.',
            picture: null
        },
        {
            id: '3',
            date: new Date().toDateString(),
            text: 'Dolor Lorem ipsum dolor sit amet consectetur adipisicing elit. Quisquam, quod.',
            picture: null
        }
    ]
})
