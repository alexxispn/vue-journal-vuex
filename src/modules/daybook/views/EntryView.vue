<template>
  <template v-if="entry">
    <div class="entry-title d-flex justify-content-between p-2">
      <div>
        <span class="text-success fs-5 fw-bold">{{ day }}</span>
        <span class="mx-1 fs-5">{{ month }}</span>
        <span class="mx-2 fw-light">{{ year }}, {{ weekday }}</span>
      </div>
      <div>
        <input v-show="false" type="file" @change="onSelectedImage" ref="fileInput" accept="image/*"/>
        <button class="btn btn-primary" @click="onUploadImage">
          Upload Image
          <i class="fas fa-cloud-upload-alt"></i>
        </button>
        <button v-if="entry.id" class="btn btn-outline-danger mx-2" @click="onDeleteEntry">
          Delete
          <i class="fas fa-trash-alt"></i>
        </button>
      </div>
    </div>
    <div class="d-flex flex-column px-3 h-75">
      <textarea v-model="entry.text" placeholder="What happened today?"></textarea>
    </div>
    <img v-if="entry.picture && !localImage" :src="entry.picture" class="img-thumbnail" alt="entry picture"/>
    <img v-if="localImage" :src="localImage" alt="entry picture" class="img-thumbnail">
  </template>
  <FabButton icon="fa-save" @on:click="saveEntry"/>
</template>

<script>
import {defineAsyncComponent} from "vue";
import {mapActions, mapGetters} from "vuex";
import Swal from "sweetalert2";
import getFullDate from "@/modules/daybook/helpers/getFullDate";
import uploadImage from "@/modules/daybook/helpers/uploadImage";


export default {
  name: "EntryView",
  props: {
    id: {
      type: String,
      required: true
    }
  },
  components: {
    FabButton: defineAsyncComponent(() => import(/* webpackChunkName: "SaveButton" */ "../components/FabButton.vue"))
  },
  data() {
    return {
      entry: null,
      localImage: null,
      file: null
    }
  },
  computed: {
    ...mapGetters('journal', ['getEntryById']),
    day() {
      const {day} = getFullDate(this.entry.date);
      return day;
    },
    month() {
      const {month} = getFullDate(this.entry.date);
      return month;
    },
    year() {
      const {year} = getFullDate(this.entry.date);
      return year;
    },
    weekday() {
      const {weekday} = getFullDate(this.entry.date);
      return weekday;
    },
  },
  methods: {
    ...mapActions('journal', ['updateEntry', 'addEntry', "deleteEntry"]),
    loadEntry() {
      let entry;
      if (this.id === 'new') {
        entry = {
          text: '',
          date: new Date().getTime()
        }
      } else {
        entry = this.getEntryById(this.id);
        if (!entry) {
          this.$router.push({name: 'no-entry'});
        }
      }
      this.entry = entry;
    },
    async saveEntry() {
      Swal.showLoading();
      this.entry.picture = await uploadImage(this.file);
      if (this.entry.id) {
        await this.updateEntry(this.entry);
      } else {
        const id = await this.addEntry(this.entry)
        this.$router.push({name: 'entry', params: {id}});
      }
      this.file = null;
      await Swal.fire('Saved', 'Entry saved successfully', 'success');
    },
    async onDeleteEntry() {
      const {isConfirmed} = await Swal.fire({
        title: 'Are you sure?',
        text: "You won't be able to revert this!",
        icon: 'warning',
        showCancelButton: true,
        confirmButtonColor: '#3085d6',
        cancelButtonColor: '#d33',
        confirmButtonText: 'Yes, delete it!'
      });
      if (isConfirmed) {
        await this.deleteEntry(this.entry.id);
        await Swal.fire(
            'Deleted!',
            'Your file has been deleted.',
            'success'
        );
        this.$router.push({name: 'no-entry'});
      }
    },
    onSelectedImage(event) {
      const image = (event.target.files[0]);
      if (!image) {
        this.localImage = null;
        this.file = null;
        return;
      }
      this.file = image;
      const fr = new FileReader();
      fr.onload = () => this.localImage = fr.result;
      fr.readAsDataURL(image);
    },
    onUploadImage() {
      this.$refs.fileInput.click();
    }
  },
  created() {
    this.loadEntry();
  },
  watch: {
    id() {
      this.loadEntry();
    }
  }
}
</script>

<style lang="scss" scoped>
textarea {
  font-size: 1.2rem;
  border: none;
  height: 100%;

  &:focus {
    outline: none;
  }
}

img {
  width: 200px;
  position: fixed;
  bottom: 150px;
  right: 20px;
  box-shadow: 0 5px 10px rgba(0, 0, 0, 0.2);
}

</style>
