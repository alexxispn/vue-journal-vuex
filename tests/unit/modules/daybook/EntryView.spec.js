import {shallowMount} from "@vue/test-utils";
import {createStore} from "vuex";
import journal from "@/modules/daybook/store/journal";
import {journalState} from "../../mock-data/journal-state";
import EntryView from "@/modules/daybook/views/EntryView";

const createVuexStore = (state) => createStore({
    modules: {
        journal: {
            ...journal,
            state
        }
    }
});
jest.mock('sweetalert2', () => ({
    fire: jest.fn(),
    showLoading: jest.fn(),
    close: jest.fn()
}))

describe('EntryView component', () => {
    const store = createVuexStore(journalState);
    store.dispatch = jest.fn();
    const mockRouter = {
        push: jest.fn()
    }
    let wrapper;
    beforeEach(() => {
        jest.clearAllMocks();
        wrapper = shallowMount(EntryView, {
            props: {
                id: '-NDTTxqbPCxX382b6UBi'
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [store]
            }
        });
    });
    it('should match snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot();
    });
    it('should redirect to /daybook if entry does not exist', () => {
        wrapper = shallowMount(EntryView, {
            props: {
                id: '123'
            },
            global: {
                mocks: {
                    $router: mockRouter
                },
                plugins: [store]
            }
        });
        expect(mockRouter.push).toHaveBeenCalledWith({name: 'no-entry'});
    });
});

