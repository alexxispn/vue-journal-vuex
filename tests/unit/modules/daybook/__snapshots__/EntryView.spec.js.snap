// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntryView component should match snapshot 1`] = `
<div class="entry-title d-flex justify-content-between p-2">
  <div><span class="text-success fs-5 fw-bold">3</span><span class="mx-1 fs-5">October</span><span class="mx-2 fw-light">2022, Monday</span></div>
  <div><input type="file" accept="image/*" style="display: none;"><button class="btn btn-primary"> Upload Image <i class="fas fa-cloud-upload-alt"></i></button><button class="btn btn-outline-danger mx-2"> Delete <i class="fas fa-trash-alt"></i></button></div>
</div>
<div class="d-flex flex-column px-3 h-75"><textarea placeholder="What happened today?"></textarea></div>
<!--v-if-->
<!--v-if-->
<fab-button-stub icon="fa-save"></fab-button-stub>
`;
